lib/recursive-methods.rb

```diff
@@ -1,49 +1,116 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n). N is the input number.
+# Space complexity: O(n). We need one stack frame for every whole number less than the input number.
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 0 || n == 1
+    return 1
+  elsif n < 0
+    raise ArgumentError
+  else
+    return n * factorial(n - 1)  
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n). N is the length of the string.
+# Space complexity: O(n). We need one stack frame for every character in the input string.
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 1 || s.length == 0
+    return s
+  else
+    return s[-1] + reverse(s[0...-1])
```

Comment:
Since doing `s[0...-1]` creates an entire new array by copying the old one over.  This **one line** is O(n) time and space complexity.